### ISSUE TITLE
Investigate validation test failure

### ISSUE THREAD

**Issue by kim** [oscoin/oscoin#387]
As encountered with c78a2c12acb48d4f1a6d08c171aa39cadc1725f6 (https://github.com/oscoin/oscoin/pull/385#issuecomment-467857584).

Build ID: 3be7950d-4a58-4dad-82a1-4c519d65d785

```
Step #6 - "Test & Build": Nakamoto: validateBlockchain
Step #6 - "Test & Build": Valid blockchains validate: FAIL (0.06s)
Step #6 - "Test & Build": *** Failed! Falsifiable (after 26 tests):
Step #6 - "Test & Build": [ETFKNPY (p:1111111) @2ms] ← [9vQCSDV (p:ETFKNPY) @1m] ← [J2GAjyj (p:9vQCSDV) @2m] ← [7wxi8VV (p:J2GAjyj) @4m] ← [93yRb3C (p:7wxi8VV) @6m] ← [8Qh9Dq2 (p:93yRb3C) @8m] ← [B79BjyP (p:8Qh9Dq2) @9m] ← [Htznjvk (p:B79BjyP) @11m] ← [Aw3zMCR (p:Htznjvk) @13m] ← [JEJk4Uv (p:Aw3zMCR) @14m] ← [2ME4WRh (p:JEJk4Uv) @15m] ← [2nUrKeo (p:2ME4WRh) @16m] ← [2vfcRpB (p:2nUrKeo) @18m] ← [CUncJSJ (p:2vfcRpB) @19m] ← [87kqY9B (p:CUncJSJ) @21m] ← [Gopsd7k (p:87kqY9B) @22m] ← [6YecCfj (p:Gopsd7k) @24m] ← [GS1oWU6 (p:6YecCfj) @26m] ← [2A2i3nW (p:GS1oWU6) @27m] ← [8TGRzfB (p:2A2i3nW) @29m] ← [5t23drn (p:8TGRzfB) @30m] ← [64vTnEz (p:5t23drn) @31m] ← [DagAXHq (p:64vTnEz) @32m] ← [13tmc6d (p:DagAXHq) @34m] ← [3r3on3S (p:13tmc6d) @35m] ← [E6r7Gdy (p:3r3on3S) @37m]
Step #6 - "Test & Build": Left (InvalidBlockDifficulty (Difficulty {fromDifficulty = 553713663}) (Difficulty {fromDifficulty = 553713663}))
Step #6 - "Test & Build": Use --quickcheck-replay=699488 to reproduce.
```

**Comment by kim** [comment]
Another counterexample:

Build ID: 0abcec97-88a7-44b2-97a1-ca92a037d279

```
Nakamoto: validateBlockchain
Valid blockchains validate: FAIL (2.81s)
*** Failed! Falsifiable (after 78 tests):
[JCVrdWg (p:1111111) @2184016h] ← [49Mx36t (p:JCVrdWg) @2184016h] ← [8ee7aeY (p:49Mx36t) @2184016h] ← [2ee6XN5 (p:8ee7aeY) @2184016h] ← [BpvjN7H (p:2ee6XN5) @2184016h] ← [DULwZf8 (p:BpvjN7H) @2184016h] ← [BY6hBD7 (p:DULwZf8) @2184016h] ← [9cuj3BH (p:BY6hBD7) @2184016h] ← [AagXdGN (p:9cuj3BH) @2184016h] ← [5ygcqBE (p:AagXdGN) @2184016h] ← [2agVWEs (p:5ygcqBE) @2184016h] ← [9q6PBMU (p:2agVWEs) @2184016h] ← [VFkdHPc (p:9q6PBMU) @2184016h] ← [GwhsAqy (p:VFkdHPc) @2184016h] ← [CDdn7EV (p:GwhsAqy) @2184016h] ← [Ed8WaXZ (p:CDdn7EV) @2184016h] ← [DZKVFpC (p:Ed8WaXZ) @2184016h] ← [EvZSesx (p:DZKVFpC) @2184016h] ← [EwhGpda (p:EvZSesx) @2184016h] ← [FJ77LZZ (p:EwhGpda) @2184016h] ← [CQJqcYc (p:FJ77LZZ) @2184016h] ← [2GCiuLe (p:CQJqcYc) @2184016h] ← [9YCGsfG (p:2GCiuLe) @2184016h] ← [LkrjC6K (p:9YCGsfG) @2184016h] ← [DP1jsin (p:LkrjC6K) @2184016h] ← [Awq5uZR (p:DP1jsin) @2184016h] ← [Cry9KLC (p:Awq5uZR) @2184016h] ← [BgWbpr1 (p:Cry9KLC) @2184016h] ← [9aMK98m (p:BgWbpr1) @2184016h] ← [2qk9biP (p:9aMK98m) @2184016h] ← [DNKdiyp (p:2qk9biP) @2184016h] ← [3LN8gjX (p:DNKdiyp) @2184016h] ← [JEJdrDf (p:3LN8gjX) @2184016h] ← [FqZVvEY (p:JEJdrDf) @2184016h] ← [FfwXDJZ (p:FqZVvEY) @2184016h] ← [BxuvkfU (p:FfwXDJZ) @2184016h] ← [C5BRvb2 (p:BxuvkfU) @2184016h] ← [7DMoWcG (p:C5BRvb2) @2184016h] ← [DWMVfnf (p:7DMoWcG) @2184017h] ← [EGz7E72 (p:DWMVfnf) @2184017h] ← [3wp6BuB (p:EGz7E72) @2184017h] ← [3NDWMf9 (p:3wp6BuB) @2184017h] ← [BSXAuj6 (p:3NDWMf9) @2184017h] ← [EdkDtik (p:BSXAuj6) @2184017h] ← [rBABYae (p:EdkDtik) @2184017h] ← [5A5MXgy (p:rBABYae) @2184017h] ← [6YXbDf4 (p:5A5MXgy) @2184017h] ← [HZiy7rH (p:6YXbDf4) @2184017h] ← [EBRiLVL (p:HZiy7rH) @2184017h] ← [AqNumMi (p:EBRiLVL) @2184017h] ← [6HC8w4K (p:AqNumMi) @2184017h] ← [4Jk32Yq (p:6HC8w4K) @2184017h] ← [4Ci84DU (p:4Jk32Yq) @2184017h] ← [4v1oWZ4 (p:4Ci84DU) @2184017h] ← [A9dWgjm (p:4v1oWZ4) @2184017h] ← [31UYs4B (p:A9dWgjm) @2184017h] ← [67Qxnbk (p:31UYs4B) @2184017h] ← [4UALBVd (p:67Qxnbk) @2184017h] ← [AZz8Lrj (p:4UALBVd) @2184017h] ← [2oYFbJj (p:AZz8Lrj) @2184017h] ← [23SfgQQ (p:2oYFbJj) @2184017h] ← [FMBrAoH (p:23SfgQQ) @2184017h] ← [Xh4Hmmu (p:FMBrAoH) @2184017h] ← [GZ8LpGp (p:Xh4Hmmu) @2184017h] ← [9N4qTMn (p:GZ8LpGp) @2184017h] ← [7wTocD4 (p:9N4qTMn) @2184017h] ← [6pNq1sG (p:7wTocD4) @2184017h] ← [7NbZxEk (p:6pNq1sG) @2184017h] ← [3wTgeHQ (p:7NbZxEk) @2184017h] ← [DRt6Kyv (p:3wTgeHQ) @2184017h] ← [DiLNQB1 (p:DRt6Kyv) @2184017h] ← [8nh7HH5 (p:DiLNQB1) @2184017h] ← [DVP32HQ (p:8nh7HH5) @2184017h] ← [6jEC1xD (p:DVP32HQ) @2184017h] ← [G93pUYG (p:6jEC1xD) @2184017h] ← [FsGh9SV (p:G93pUYG) @2184017h] ← [GFs5w5T (p:FsGh9SV) @2184017h] ← [GsbQBU3 (p:GFs5w5T) @2184017h]
Left (InvalidBlockDifficulty (Difficulty {fromDifficulty = 553713663}) (Difficulty {fromDifficulty = 553713663}))
Use --quickcheck-replay=736411 to reproduce.
```

**Comment by adinapoli-mndc** [comment]
@kim @cloudhead I bet this is introduced by the generators I wrote as part of #373 and #320 : now the generator for `Difficulty` can actually "swing back" to simulate network adjustments, and my 9AM-in-the-morning speculation is that our current validators wants the difficulty to always be monotonically increasing with respect to the previous block. Would that be plausible? 🤔

**Comment by kim** [comment]
@adinapoli-mndc Looks about right:

https://github.com/oscoin/oscoin/blob/master/src/Oscoin/Consensus/Nakamoto.hs#L91-L93

Definition of `hasPoW`:

```haskell
hasPoW header = difficulty header < blockTargetDifficulty header
```

**Comment by cloudhead** [comment]
Not sure - the counter example shows the blocks having the same difficulty?

**Comment by cloudhead** [comment]
Nevermind I was looking at the wrong thing!

**Comment by cloudhead** [comment]
The problem could be that the generator is creating difficulty swings, but the block seal is arbitrary, so it may or may not be `<` than the target difficulty. Doesn't explain though how it matches exactly.

**Comment by cloudhead** [comment]
> always be monotonically increasing with respect to the previous block. Would that be plausible?

This I don't think is the case - we compute the target based on previous blocks, but there's no requirement for it to increase.

**Comment by kim** [comment]
@cloudhead The target difficulty is _generated_ from the previous, with a 20% chance of it being the same https://github.com/oscoin/oscoin/blob/f1e588220644e93e8a611198eece461c2ba5af19/test/Oscoin/Test/Crypto/Blockchain/Block/Generators.hs#L54

**Comment by adinapoli-mndc** [comment]
And mind, I did write that generator, so it's highly possible this is nonsense 😁

**Comment by cloudhead** [comment]
>  @cloudhead The target difficulty is _generated_ from the previous, with a 20% chance of it being the same

I get this - but my understanding is that the problem lies in the actual difficulty of the block, not the target - since the validation of the target seems to have passed. The only thing that affects the actual difficulty is the hash, which we don't look at when generating (AFAICT)

**Comment by kim** [comment]
Ah right. I guess this is then due to the compact representation of difficulty not being lossless.

Definition of `difficulty`:

```haskell
difficulty = encodeDifficulty . os2ip . headerHash
```

There is a wide range of `Integer`s which map to the same `Difficulty`. It is also the case that `553713663` happens to be the `minDifficulty`. I guess the shrinking of the block header inputs in the generator will need some tweaking.

**Comment by cloudhead** [comment]
Aha, yeah looks like that's it!

**Comment by adinapoli-mndc** [comment]
I have tried to reproduce this issue with the current HEAD and I cannot reproduce it anymore 😲 , neither with the mockcrypto not with the real world one. I have tried both seeds reported in the issue and none of them yielded a failure 🤔 

I am not sure what happened, but that seemed to be a transient issue. I have even tried to checkout an old commit (149cec64eda4e000844d09a4105687b536b4a51a) but no dice, no failure to be seen.

We suggest we close this ticket and we keep an eye on it.
